### PR TITLE
DataArray plot namespace attribute

### PR DESCRIPTION
**Pull request by clarkfitzg**
Use `DataArray.plot` attribute to access `xray.plotting` functions as methods on a DataArray.

Example: `da.plot_imshow()` becomes `da.plot.imshow()`, which is equivalent to `xray.plotting.imshow(da)`.

Not finished yet, but putting it up here for visibility and feedback.

### CODE REVIEW

File: doc/plotting.rst [hunk 1]
```diff
@@ -66,7 +66,7 @@ Additional Arguments
 
 Additional arguments are passed directly to the matplotlib function which
 does the work.
-For example, :py:meth:`xray.DataArray.plot_line` calls 
+For example, :py:meth:`xray.DataArray.plot.line` calls
```
Comment (clarkfitzg) [inline doc/plotting.rst]: Thinking to just link these to `xray.plotting.line` function. Right now they don't work.

File: xray/plotting/plotting.py [hunk 2]
```diff
@@ -441,3 +441,69 @@ def plot_pcolormesh(x, y, z, ax, **kwargs):
...
 
     return ax, primitive
+
+
+class _PlotMethods(object):
+    '''
+    Plotmethods class
+    '''
+
+    # Basically this is a bunch of hacking to put in the dataarray as the
```
Comment (clarkfitzg) [inline xray/plotting/plotting.py]: The current implementation works, but I'm not happy with it. The biggest problem is the way all the 2d plots are added- they all have the same form, so should be added in a more DRY way.

File: xray/plotting/plotting.py [hunk 3]
```diff
@@ -61,10 +61,10 @@ def plot(darray, ax=None, rtol=0.01, **kwargs):
     =============== =========== ===========================
```
Comment (clarkfitzg) [inline xray/plotting/plotting.py]: Maybe better to rename this `plot/plot` to be consistent with `DataArray.plot`. Thoughts?

File: xray/plotting/plotting.py [hunk 4]
```diff
@@ -441,3 +441,69 @@ def plot_pcolormesh(x, y, z, ax, **kwargs):
...
+    @functools.wraps(hist)
+    def hist(self, ax=None, **kwargs):
+        return hist(self._da, ax=ax, **kwargs)
+
+    @functools.wraps(line)
+    def line(self, *args, **kwargs):
+        return line(self._da, *args, **kwargs)
+
+    @functools.wraps(plot)
+    def plot(darray, ax=None, rtol=0.01, **kwargs):
```
Comment (shoyer) [inline xray/plotting/plotting.py]: does it make sense to include this method separately?

File: xray/plotting/plotting.py [hunk 5]
```diff
@@ -441,3 +441,69 @@ def plot_pcolormesh(x, y, z, ax, **kwargs):
...
+        #for f in _plot2dlist:
+        #    setattr(self, f.__name__, self._2dplot(f))
+
+    def __call__(self, *args, **kwargs):
+        return plot(self._da, *args, **kwargs)
+
+    def _2dplot(self, plotfunc):
+        @functools.wraps(plotfunc)
+        def inner(self, ax=None, xincrease=None, yincrease=None,
+                  add_colorbar=True, **kwargs):
```
Comment (shoyer) [inline xray/plotting/plotting.py]: don't forget the new arguments I added to the 2d plotting signatures

File: xray/plotting/plotting.py (hunk 3 above)
Comment (shoyer) [inline xray/plotting/plotting.py]: I'm not quite sure I understand -- discuss this offline?

File: xray/plotting/plotting.py (hunk 4 above)
Comment (clarkfitzg) [inline xray/plotting/plotting.py]: Ah, we can already just call plot directly. So this method can be removed.

File: xray/plot/plot.py [hunk 6]
```diff
@@ -257,9 +260,34 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
     return vmin, vmax, cmap, extend
 
 
+# MUST run before any 2d plotting functions are defined since
+# _plot2d decorator adds them as methods here.
```
Comment (clarkfitzg) [inline xray/plot/plot.py]: This is the best solution I've come up with for being able to use `DataArray.plot.<plotfunc>`. If there's a better way to do it I'm all ears.

File: xray/plot/plot.py [hunk 7]
```diff
@@ -61,10 +64,10 @@ def plot(darray, ax=None, rtol=0.01, **kwargs):
     =============== =========== ===========================
     Dimensions      Coordinates Plotting function
     --------------- ----------- ---------------------------
-    1                           :py:meth:`xray.DataArray.plot_line`
-    2               Uniform     :py:meth:`xray.DataArray.plot_imshow`
-    2               Irregular   :py:meth:`xray.DataArray.plot_contourf`
-    Anything else               :py:meth:`xray.DataArray.plot_hist`
+    1                           :py:meth:`xray.DataArray.line`
```
Comment (shoyer) [inline xray/plot/plot.py]: these should be `xray.plot.line`, right?

**Comment by shoyer** [comment]
This looks pretty reasonable to me. Do the docs generate properly?

**Comment by clarkfitzg** [comment]
No not locally. Getting an error on this:
```
     50     # Not sure what the first arg here should be
---> 51     f = pkg_resources.resource_stream(__name__, fname)
```
Which means it's time for me to actually understand this part :)

**Comment by clarkfitzg** [comment]
Got it- was the name of the file in setup.py again.

**Comment by clarkfitzg** [comment]
I'm happy with this as it stands. Although in the future I would probably not try to make an attribute behave like a namespace. Too awkward.

**Comment by shoyer** [comment]
OK, I'm going to merge this

**Comment by clarkfitzg** [comment]
go for it!

**Pull request merged**